### PR TITLE
Update ILL reflectometry loader and D17 and Figaro IDF

### LINKED ISSUE #18235: Update LoadILLReflectometry
### Expected behavior

Update loading of D17 and Figaro Nexus file data:

## Initial
- add error values to monitors
- use variable `m_supportedInstruments`
- convert TOF units to wavelength
- if TOF axis cannot be defined, still be able to load the data

## Scattering angle, coherent and incoherent scattering options 
- load direct beam file if required by coherent scattering,  detector angle
- correct positioning detectors
- various options to calculate the scattering angle (angle, coherence)
- input validation

## Add instrument Figaro
- load Figaro Nexus files and enable options like in D17

## Changes of the behavior of the loader
- no SampleLog entry channel_width, since this value is already present (PSD.time_of_flight_0). This was not documented before, but will be documented now.
- corrected rotation of the detector (via calculation of the scattering angle)

Those changes do not justify to create a version 2 of the loader in my opinion. 

### Platforms affected

All

### PR DESCRIPTION
**Pull request by soininen**
This PR updates [`LoadILLReflectometry`](http://docs.mantidproject.org/nightly/algorithms/LoadILLReflectometry-v1.html) to correctly load current D17 NeXus files. The file format for the Figaro instrument is still under development so the loader cannot open those files yet.

Additionally, the IDFs of D17 and Figaro were updated.

**To test:**

The output of the loader has been validated against the legacy LAMP software at ILL. Therefore a code review and the standard try-to-break-it should suffice. A test script might look like this (add your Mantid build path to the appropriate position):
```python
mantid_data_dir = '<tester input needed here>/ExternalData/Testing/Data/UnitTest/ILL/D17/'
mantid.config.appendDataSearchDir(mantid_data_dir)

# Load direct beam data, note the extra optional output
LoadILLReflectometry(Filename='317369', OutputWorkspace='DB', OutputBeamPosition='beam_position', XUnit='TimeOfFlight')
# Load reflected beam. Use direct beam information to calibrate the detector angle.
LoadILLReflectometry(Filename='317370', OutputWorkspace='RB', BeamPosition='beam_position', XUnit='TimeOfFlight')
```

Fixes #18235 and #19029.

**Release Notes** 

The release notes were updated accordingly.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by NickDraper** [comment]
@soininen  Could you revsolve the conflicts here

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 1]
```diff
@@ -1,22 +1,181 @@
...
+
+namespace {
+/// A struct for information needed for detector angle calibration.
+struct DirectBeamMeasurement {
+  double detectorAngle;
+  double detectorDistance;
+  double fittedPeakCentre;
+  double positionOfMaximum;
+};
+
```
Comment (josephframsay) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: Could any of the helper functions below also go into the anonymous namespace?

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 2]
```diff
@@ -62,406 +197,617 @@ int LoadILLReflectometry::confidence(
...
+    m_detectorDistance = "det";
+    m_detectorAngleName = "dan.value";
+    m_sampleAngleName = "san.value";
+    m_offsetFrom = "VirtualChopper";
+    m_offsetName = "open_offset";
+    m_pixelCentre = 127.5;
+    m_chopper1Name = "Chopper1";
+    m_chopper2Name = "Chopper2";
+  } else if (m_instrumentName == "Figaro") {
+    // TODO Figaro's detector position should be calculated from
```
Comment (josephframsay) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: If these TODOs are done, could they be removed?

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 3]
```diff
@@ -62,406 +197,617 @@ int LoadILLReflectometry::confidence(
...
-  //      pos.getSpherical(r, theta, phi);
-
-  double angle_rad = angle * deg2rad;
-  V3D newpos(distance * sin(angle_rad), pos.Y(), distance * cos(angle_rad));
+/// Update detector position according to data file
+void LoadILLReflectometry::placeDetector() {
+  g_log.debug("Move the detector bank \n");
+  double dist = doubleFromRun(m_detectorDistance + ".value");
+  m_detectorDistanceValue = inMeter(dist);
+  // TODO offset_value cannot be used like this for Figaro.
```
Comment (josephframsay) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: Again are these TODOs still to do, or should be they be removed?

**Review by josephframsay — CHANGES_REQUESTED** [review]
Changes look good, and testing works as expected. A few minor points of style, but other than that everything looks cool

File: Framework/DataHandling/src/LoadILLReflectometry.cpp (hunk 1 above)
Comment (soininen) [reply Framework/DataHandling/src/LoadILLReflectometry.cpp]: Which helper functions? The anonymous namespace continues all the way to line 178 and should include all the functions 'below'. Did you mean some private method of `LoadILLReflectometry`? I think all of them use a member variable or the logger but maybe I missed something?

Comment (josephframsay) [reply Framework/DataHandling/src/LoadILLReflectometry.cpp]: Oh whoops, my bad - got my curly braces mixed up. I presumably meant all the helper functions already in the anonymous namespace

File: Framework/DataHandling/src/LoadILLReflectometry.cpp (hunk 2 above)
Comment (soininen) [reply Framework/DataHandling/src/LoadILLReflectometry.cpp]: This loader loads current Figaro files incorrectly, and the TODOs are there to remind us about it. In fact, the file formats for both reflectometers at ILL will change in the near future, and we already have plans for a large overhaul of the loader, and an update to version 2. Meanwhile, however, we need something that can load D17 files. This enables us to build the data reduction workflow for the reflectometers.

Comment (josephframsay) [reply Framework/DataHandling/src/LoadILLReflectometry.cpp]: Yup, I see. In that case, In that case everything looks good!

**Review by josephframsay — APPROVED** [review]

**Pull request merged**